### ISSUE TITLE
Type coercions in the JOIN ON clause broken 

### ISSUE THREAD

**Issue by nileema** [facebook/presto#2419]
The following query fails: 
```
SELECT *
FROM
 (VALUES 1.0) a(x)
 JOIN 
 (VALUES 2) b
 ON a.x = 1
```

**Comment by yuananf** [comment]
Got a similar problem here.

Two tables a1 and a2 has the exact same types and values

presto:test> desc a1;
 Column |  Type  | Null | Partition Key | Comment 
--------+--------+------+---------------+---------
 t1     | bigint | true | false         |         
 t2     | double | true | false         |

presto:test> desc a2;
 Column |  Type  | Null | Partition Key | Comment 
--------+--------+------+---------------+---------
 t1     | bigint | true | false         |         
 t2     | double | true | false         |

presto:test> select * from a1;
 t1 | t2  
----+-----
  1 | 1.0 
  2 | 2.0

presto:test> select * from a2;
 t1 | t2  
----+-----
  1 | 1.0 
  2 | 2.0

simple inner join got nothing return.
presto:test> select a1.* from a1 join a2 on a1.t1 = a2.t2;
 t1 | t2 
----+----
(0 rows)

implicit join got the correct result.
presto:test> select a1.* from a1, a2 where a1.t1 = a2.t2;
 t1 | t2  
----+-----
  1 | 1.0 
  2 | 2.0

Can anyone explain this result?
@cberner

**Comment by yuananf** [comment]
Fixed by #2542
